File: GPT和Gemini生图带压测/tasks.py
"""异步任务队列:批量生图任务排队、顺序执行、可查询进度。

任务完成后写入历史(store.add_history)。仅文生图走队列;
图生图(带文件上传)由 app 直接同步调用。
"""
import asyncio
import time
import uuid
from collections import OrderedDict

import config
import generator
import store

# 任务状态: queued -> running -> done | error
_tasks: "OrderedDict[str, dict]" = OrderedDict()
_queue: asyncio.Queue | None = None
_worker_started = False
_running_workers = 0  # 当前实际运行的 worker 数(并发数)
MAX_KEEP = 100  # 内存中最多保留的任务记录数


def running_workers() -> int:
    return _running_workers


def _now() -> int:
    return int(time.time())


def enqueue(*, prompt, size, quality, n, model=None, request_format="images") -> dict:
    """加入一个文生图任务,返回任务记录。"""
    task = {
        "id": uuid.uuid4().hex[:12],
        "mode": "gen",
        "status": "queued",
        "prompt": prompt,
        "size": size,
        "quality": quality,
        "n": n,
        "model": model,
        "request_format": request_format,
        "created_at": _now(),
        "started_at": None,
        "finished_at": None,
        "images": [],
        "error": None,
    }
    _tasks[task["id"]] = task
    # 超出上限时清理最旧的已结束任务
    while len(_tasks) > MAX_KEEP:
        for tid, t in list(_tasks.items()):
            if t["status"] in ("done", "error"):
                _tasks.pop(tid, None)
                break
        else:
            break
    assert _queue is not None, "队列未启动"
    _queue.put_nowait(task["id"])
    return task


def get(task_id: str) -> dict | None:
    return _tasks.get(task_id)


def list_tasks(limit: int = 50) -> list:
    items = list(_tasks.values())[-limit:]
    return list(reversed(items))


async def _run_one(task_id: str):
    task = _tasks.get(task_id)
    if not task:
        return
    task["status"] = "running"
    task["started_at"] = _now()
    cfg = config.load()
    fmt = task.get("request_format") or cfg.get("request_format", "images")
    try:
        common = dict(
            prompt=task["prompt"],
            base_url=cfg["base_url"],
            api_key=cfg["api_key"],
            model=task["model"] or cfg["model"],
            size=task["size"] or cfg["default_size"],
            quality=task["quality"] or cfg["default_quality"],
            n=task["n"],
            timeout=int(cfg.get("timeout", 300)),
        )
        if fmt == "chat":
            results = await generator.generate_via_chat(**common)
        else:
            results = await generator.generate(**common)
        task["images"] = [{"filename": r["filename"], "url": r["url"]} for r in results]
        task["status"] = "done"
        store.add_history(
            mode="gen", prompt=task["prompt"], model=task["model"] or cfg["model"],
            size=task["size"], quality=task["quality"], n=task["n"],
            files=[r["filename"] for r in results],
        )
    except generator.GenerationError as e:
        task["status"] = "error"
        task["error"] = str(e)
    except Exception as e:  # 兜底,避免 worker 因单个任务崩溃
        task["status"] = "error"
        task["error"] = f"内部错误:{e}"
    finally:
        task["finished_at"] = _now()


async def _worker(worker_id: int):
    while True:
        task_id = await _queue.get()
        try:
            await _run_one(task_id)
        finally:
            _queue.task_done()


def start():
    """在 FastAPI startup 时调用,初始化队列并按并发数启动多个 worker。"""
    global _queue, _worker_started, _running_workers
    if _worker_started:
        return
    _queue = asyncio.Queue()
    concurrency = max(1, min(int(config.load().get("concurrency", 3)), 16))
    for i in range(concurrency):
        asyncio.create_task(_worker(i))
    _running_workers = concurrency
    _worker_started = True

File: internal/tasks/tasks.go
// Package tasks 实现异步任务队列:批量文生图任务排队、并发执行、可查询进度。
// 任务完成后写入历史(store.AddHistory)。图生图(带文件上传)由 server 直接同步调用,不走队列。
package tasks

import (
	"context"
	"crypto/rand"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/meihai66/stcs/internal/config"
	"github.com/meihai66/stcs/internal/generator"
	"github.com/meihai66/stcs/internal/store"
)

const maxKeep = 100

// Task 是一个文生图任务。
type Task struct {
	ID            string           `json:"id"`
	UserID        int64            `json:"user_id"`
	Username      string           `json:"username,omitempty"` // 仅管理员「查看全部」时填充
	Mode          string           `json:"mode"`
	Status        string           `json:"status"` // queued|running|done|error
	Prompt        string           `json:"prompt"`
	Size          string           `json:"size"`
	Quality       string           `json:"quality"`
	N             int              `json:"n"`
	Model         string           `json:"model"`
	RequestFormat string           `json:"request_format"`
	CreatedAt     int64            `json:"created_at"`
	StartedAt     *int64           `json:"started_at"`
	FinishedAt    *int64           `json:"finished_at"`
	Images        []store.Image    `json:"images"`
	Error         string           `json:"error"`
}

var (
	mu       sync.Mutex
	tasks    = map[string]*Task{}
	order    []string // 保持插入顺序
	queue    chan string
	started  bool
	workers  int
)

// Start 初始化队列并按并发数启动 worker。幂等。
func Start() {
	mu.Lock()
	defer mu.Unlock()
	if started {
		return
	}
	c := config.GlobalConcurrency()
	if c < 1 {
		c = 1
	}
	if c > 16 {
		c = 16
	}
	workers = c
	queue = make(chan string, 4096)
	for i := 0; i < c; i++ {
		go worker()
	}
	started = true
}

// RunningWorkers 返回当前 worker 数。
func RunningWorkers() int {
	mu.Lock()
	defer mu.Unlock()
	return workers
}

func now() int64 { return time.Now().Unix() }

func randID() string {
	b := make([]byte, 6)
	_, _ = rand.Read(b)
	return fmt.Sprintf("%x", b)
}

// Enqueue 加入一个文生图任务,返回任务快照。
func Enqueue(userID int64, prompt, size, quality, model, requestFormat string, n int) Task {
	mu.Lock()
	t := &Task{
		ID:            randID(),
		UserID:        userID,
		Mode:          "gen",
		Status:        "queued",
		Prompt:        prompt,
		Size:          size,
		Quality:       quality,
		N:             n,
		Model:         model,
		RequestFormat: requestFormat,
		CreatedAt:     now(),
		Images:        []store.Image{},
	}
	tasks[t.ID] = t
	order = append(order, t.ID)
	// 超出上限清理最旧的已结束任务
	for len(order) > maxKeep {
		removed := false
		for i, id := range order {
			if tt := tasks[id]; tt != nil && (tt.Status == "done" || tt.Status == "error") {
				delete(tasks, id)
				order = append(order[:i], order[i+1:]...)
				removed = true
				break
			}
		}
		if !removed {
			break
		}
	}
	snap := *t
	mu.Unlock()
	queue <- t.ID
	return snap
}

// Get 返回一个任务快照。
func Get(id string) (Task, bool) {
	mu.Lock()
	defer mu.Unlock()
	t, ok := tasks[id]
	if !ok {
		return Task{}, false
	}
	return *t, true
}

// List 倒序返回最近的任务快照。all=true(管理员)返回全部用户,否则只返回 userID 的。
func List(userID int64, all bool, limit int) []Task {
	mu.Lock()
	defer mu.Unlock()
	out := make([]Task, 0, limit)
	for i := len(order) - 1; i >= 0 && len(out) < limit; i-- {
		t := tasks[order[i]]
		if t == nil || (!all && t.UserID != userID) {
			continue
		}
		out = append(out, *t)
	}
	return out
}

// Clear 清理已结束(done/error)的任务,保留排队中/生成中的,返回清理数量。
// all=true(管理员)清全部用户,否则只清 userID 的。
func Clear(userID int64, all bool) int {
	mu.Lock()
	defer mu.Unlock()
	kept := make([]string, 0, len(order))
	removed := 0
	for _, id := range order {
		t := tasks[id]
		if t != nil && (all || t.UserID == userID) && (t.Status == "done" || t.Status == "error") {
			delete(tasks, id)
			removed++
			continue
		}
		kept = append(kept, id)
	}
	order = kept
	return removed
}

func worker() {
	for id := range queue {
		runOne(id)
	}
}

func setStatus(id, status, errMsg string, images []store.Image, started, finished bool) {
	mu.Lock()
	defer mu.Unlock()
	t := tasks[id]
	if t == nil {
		return
	}
	if status != "" {
		t.Status = status
	}
	if errMsg != "" {
		t.Error = errMsg
	}
	if images != nil {
		t.Images = images
	}
	if started {
		n := now()
		t.StartedAt = &n
	}
	if finished {
		n := now()
		t.FinishedAt = &n
	}
}

func runOne(id string) {
	mu.Lock()
	t := tasks[id]
	if t == nil {
		mu.Unlock()
		return
	}
	snap := *t
	mu.Unlock()

	setStatus(id, "running", "", nil, true, false)

	cfg := config.LoadForUser(snap.UserID)
	fmtMode := snap.RequestFormat
	if fmtMode == "" {
		fmtMode = cfg.RequestFormat
	}
	model := snap.Model
	if model == "" {
		model = cfg.Model
	}
	size := snap.Size
	if size == "" {
		size = cfg.DefaultSize
	}
	quality := snap.Quality
	if quality == "" {
		quality = cfg.DefaultQuality
	}
	p := generator.Params{
		UserID:  snap.UserID,
		Prompt:  snap.Prompt,
		BaseURL: cfg.BaseURL,
		APIKey:  cfg.APIKey,
		Model:   model,
		Size:    size,
		Quality: quality,
		N:       snap.N,
		Timeout: cfg.Timeout,
	}

	var results []generator.Result
	var err error
	if fmtMode == "chat" {
		results, err = generator.GenerateViaChat(context.Background(), p)
	} else {
		results, err = generator.Generate(context.Background(), p)
	}
	if err != nil {
		msg := err.Error()
		if !generator.IsGenError(err) {
			msg = "内部错误:" + msg
		}
		setStatus(id, "error", msg, nil, false, true)
		return
	}
	imgs := make([]store.Image, 0, len(results))
	files := make([]string, 0, len(results))
	for _, r := range results {
		imgs = append(imgs, store.Image{Filename: r.Filename, URL: r.URL})
		files = append(files, r.Filename)
	}
	setStatus(id, "done", "", imgs, false, true)
	store.AddHistory(snap.UserID, "gen", snap.Prompt, model, size, quality, snap.N, files)
	pruneUserImages(snap.UserID)
}

// pruneUserImages 按用户图片上限裁掉最旧历史,并从磁盘删除对应文件。
func pruneUserImages(userID int64) {
	u, ok := store.GetUser(userID)
	if !ok || u.ImageLimit <= 0 {
		return
	}
	for _, f := range store.PruneUserImages(userID, u.ImageLimit) {
		_ = os.Remove(filepath.Join(generator.UserDir(userID), filepath.Base(f)))
	}
}
